x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/TRange.java

```diff
@@ -0,0 +1,363 @@
+                .and(
+                    isType(
+                        first,
+                        dt -> isTemporalAmount(dt) || dt == KEYWORD || dt == LONG,
+                        operationName,
+                        FIRST,
+                        "time_duration",
+                        "date_period",
+                        "string",
+                        "long"
+                    )
+                );
+        }
+
+        // Two parameter mode
+        resolution = isNotNull(first, operationName, FIRST).and(isFoldable(first, operationName, FIRST))
+            .and(isNotNull(second, operationName, SECOND))
+            .and(isFoldable(second, operationName, SECOND));
+
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        // the 2nd parameter is an absolute time in string or long format: (<string>, <string>) or (<long>, <long>)
+        if (isTemporalAmount(second.dataType()) == false) {
+            resolution = isType(first, dt -> dt == KEYWORD || dt == LONG, operationName, FIRST, "string", "long").and(
+                isType(second, dt -> dt == first.dataType(), operationName, SECOND, first.dataType().esType())
+            );
+
+            if (resolution.unresolved()) {
+                return resolution;
+            }
+        } else {
+            // the 2nd parameter is a temporal amount (offset): (<string|long>, <time_duration|date_period>)
+            resolution = isType(first, dt -> dt == KEYWORD || dt == LONG, operationName, FIRST, "string", "long").and(
+                isType(second, DataType::isTemporalAmount, operationName, SECOND, "time_duration", "date_period")
+            );
+
+            if (resolution.unresolved()) {
+                return resolution;
+            }
+        }
+
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        if (newChildren.size() == 3) {
+            return new TRange(source(), newChildren.get(0), newChildren.get(1), newChildren.get(2), configuration());
+        }
+        return new TRange(source(), newChildren.get(0), newChildren.get(1), null, configuration());
+    }
+
+    @Override
+    public Expression surrogate() {
+        long[] range = getRange(FoldContext.small());
+
+        Expression startLiteral;
+        Expression endLiteral;
+
+        if (timestamp.dataType() == DataType.DATE_NANOS) {
+            startLiteral = new Literal(source(), range[0], DataType.DATE_NANOS);
+            endLiteral = new Literal(source(), range[1], DataType.DATE_NANOS);
+        } else {
+            startLiteral = new Literal(source(), range[0], DataType.DATETIME);
+            endLiteral = new Literal(source(), range[1], DataType.DATETIME);
+        }
+
+        return new And(
+            source(),
+            new GreaterThanOrEqual(source(), timestamp, startLiteral),
+            new LessThanOrEqual(source(), timestamp, endLiteral)
```

Comment:
Should this be LessThan instead? I think PromQL range selection is left-open and right-closed.